ext/ddtrace_profiling_native_extension/collectors_cpu_and_wall_time.c

```diff
@@ -238,6 +292,190 @@ VALUE cpu_and_wall_time_collector_sample(VALUE self_instance) {
+    ) continue; // Ignore threads with no/incomplete garbage collection data
+
+    sampled_any_thread = true;
+
+    long gc_cpu_time_elapsed_ns =
+      thread_context->gc_tracking.cpu_time_at_finish_ns - thread_context->gc_tracking.cpu_time_at_start_ns;
+    long gc_wall_time_elapsed_ns =
+      thread_context->gc_tracking.wall_time_at_finish_ns - thread_context->gc_tracking.wall_time_at_start_ns;
+
+    if (gc_cpu_time_elapsed_ns < 0) rb_raise(rb_eRuntimeError, "BUG: Unexpected negative gc_cpu_time_elapsed_ns between samples");
+    if (gc_wall_time_elapsed_ns < 0) rb_raise(rb_eRuntimeError, "BUG: Unexpected negative gc_wall_time_elapsed_ns between samples");
+
+    if (thread_context->gc_tracking.wall_time_at_start_ns == 0 && thread_context->gc_tracking.wall_time_at_finish_ns != 0) {
+      // Avoid using wall-clock if we got 0 for a start (meaning there was an error) but not 0 for finish so we don't
+      // come up with a crazy value for the frame
+      rb_raise(rb_eRuntimeError, "BUG: Unexpected zero value for gc_tracking.wall_time_at_start_ns");
+    }
+
+    int64_t metric_values[ENABLED_VALUE_TYPES_COUNT] = {0};
+
+    metric_values[CPU_TIME_VALUE_POS] = gc_cpu_time_elapsed_ns;
+    metric_values[CPU_SAMPLES_VALUE_POS] = 1;
+    metric_values[WALL_TIME_VALUE_POS] = gc_wall_time_elapsed_ns;
+
+    trigger_sample_for_thread(
+      state,
+      thread,
+      thread_context,
+      (ddog_Slice_i64) {.ptr = metric_values, .len = ENABLED_VALUE_TYPES_COUNT},
+      SAMPLE_IN_GC
+    );
+
+    // Mark thread as no longer in GC
+    thread_context->gc_tracking.cpu_time_at_start_ns = INVALID_TIME;
+    thread_context->gc_tracking.cpu_time_at_finish_ns = INVALID_TIME;
+    thread_context->gc_tracking.wall_time_at_start_ns = INVALID_TIME;
+    thread_context->gc_tracking.wall_time_at_finish_ns = INVALID_TIME;
+
+    // Update counters so that they won't include the time in GC during the next sample
+    if (thread_context->cpu_time_at_previous_sample_ns != INVALID_TIME) {
+      thread_context->cpu_time_at_previous_sample_ns += gc_cpu_time_elapsed_ns;
+    }
+    if (thread_context->wall_time_at_previous_sample_ns != INVALID_TIME) {
+      thread_context->wall_time_at_previous_sample_ns += gc_wall_time_elapsed_ns;
+    }
+  }
+
+  if (sampled_any_thread) state->stats.gc_samples++;
+
+  // Return a VALUE to make it easier to call this function from Ruby APIs that expect a return value (such as rb_rescue2)
+  return Qnil;
+}
+
+static void trigger_sample_for_thread(
+  struct cpu_and_wall_time_collector_state *state,
+  VALUE thread,
+  struct per_thread_context *thread_context,
+  ddog_Slice_i64 metric_values_slice,
+  sample_type type
+) {
+  VALUE thread_name = thread_name_for(thread);
+  bool have_thread_name = thread_name != Qnil;
+
+  int label_count = 1 + (have_thread_name ? 1 : 0);
+  ddog_Label labels[label_count];
+
+  labels[0] = (ddog_Label) {.key = DDOG_CHARSLICE_C("thread id"), .str = thread_context->thread_id_char_slice};
```

Comment:
A bit unrelated, but where's the definition of `DDOG_CHARSLICE_C`? I can only find the definition of `VALUE_STRING`.

Reply:
It's in `libdatadog/common.h`.

It gets set up here https://github.com/DataDog/libdatadog/blob/81b1753db3454c39f96446cae0c7397ded032a79/ddcommon-ffi/cbindgen.toml#L17-L23 and you can see the generated file by runnning `bundle open libdatadog` and finding the `common.h` file.

Reply:
Thank you!